### PR TITLE
chore: update PR template for contributions (#3214)

### LINKED ISSUE #3214: Discussion for enriching the Github templates
We'd like to enrich the existing PR template adding a few sections that we consider helpful during the review process for a PR, making it even better.

>One **important** thing for us is not to overwhelm whoever is filling out this template, so please feel free to discuss about removing parts of it if that's the case. In fact we received very positive feedback from the Java team to adapt to their needs. The same applies here 😄 

In this template, besides the checklist that we already have, we'd like to define the following structure, where:
- each section is a H2 element
- the mandatory/recommended/optional keyword goes in an HTML comment (so won't appear in the end description)
- **the description texts go in an HTML comment** (so won't appear in the end description)

You could see examples for [Beats](https://github.com/elastic/beats/pull/15388), and [APM Agent Java](https://github.com/elastic/apm-agent-java/pull/994)

We will also use this markdown format (H2 and HTML comments) for the other templates you already have.

The initial template: ⬇️⬇️⬇️⬇️ 

A few suggestions about filling out this PR

1. Use a descriptive title for the PR.
2. If this pull request is work in progress, create a draft PR instead of prefixing the title with WIP.
3. Please label this PR at least one of the following labels, depending on the scope of your change:
- feature request, which adds new behavior
- bug fix
- enhancement, which modifies existing behavior
- breaking change
4. Remove those recommended/optional sections if you don't need them. Only "What does this PR do" and "Checklist" are mandatory.
5. Submit the pull request: Push your local changes to your forked copy of the repository and submit a pull request (https://help.github.com/articles/using-pull-requests).
6. Please be patient. We might not be able to review your code as fast as we would like to, but we'll do our best to dedicate to it the attention it deserves. Your effort is much appreciated!

>Do you see any other type?

## What does this PR do? _(Mandatory)_
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.

## Why is it important? _(Mandatory)_
Explain here the WHY, or the rationale/motivation for the changes.

## Checklist _(Mandatory)_
Add a checklist of things that are required to be reviewed in order to have the PR approved
List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply.

>Here we have a default list, based on Beats and Java/Ruby agents:

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/). 
- [ ] My code follows the style guidelines of this project
- [ ] I have rebased my changes on top of the latest master branch
Update your local repository with the most recent code from the main repo, and rebase your branch on top of the latest master branch. We prefer your initial changes to be squashed into a single commit. Later, if we ask you to make changes, add them as separate commits. This makes them easier to review.
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally with my changes
Run the test suite to make sure that nothing is broken. See https://github.com/elastic/apm-server/blob/master/TESTING.md for details.
- [ ] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)

## Author's Checklist _(Recommended)_
Add a checklist of things that are required to be reviewed in order to have the PR approved

## How to test this PR locally _(Recommended)_
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.

## Related issues _(Recommended)_
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.
- Closes #ISSUE_ID
- Relates #ISSUE_ID
- Requires #ISSUE_ID
- Superseds #ISSUE_ID

## Use cases _(Recommended)_
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.
If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/

## Logs _(Recommended)_
Paste here output logs discovered while creating this PR, or any other output you consider important to be shared with the team. It could be a stacktrace, logs of a library...

## Screenshots _(Optional)_
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.

### PR DESCRIPTION
**Pull request by mdelapenya**
## Motivation/summary
We want to improve the experience of contributing to this repo, for that reason we created a more structured format for the existing template, moving some bullet points from the CONTRIBUTING guide to a checklist that will be present on each PR submission.

The format will consist in H2 headers and HTML comments, apart from a checklist for items that must be addressed by sender & reviewer during the review process.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/). 
- [x] My code follows the style guidelines of this project
- [x] I have rebased my changes on top of the latest master branch

- [x] I have made corresponding changes to the documentation
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] New and existing [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally with my changes

~~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~~

## Related issues

Closes #3214

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -21,51 +21,6 @@ or that there are particular issues that you should know about before implementi
 You will have to fork the `apm-server` repo,
 please follow the instructions in the [readme](README.md).
```
Comment (kuisathaverat) [inline CONTRIBUTING.md]: removing this section creates a gab in the contributing document, it is true that the info in the template, but if I read the contributing document there are some steps that I will only see when the PR is ready.

```suggestion
  [Checks before submitting your changes](.github/pull_request_template.md)
```

Comment (mdelapenya) [reply CONTRIBUTING.md]: I like this approach! @simitt @axw wdyt?

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -27,3 +31,41 @@ Major changes require a number of considerations including impact on:
 * Elasticsearch Service (https://cloud.elastic.co)
 * Elastic Cloud Enterprise (https://www.elastic.co/products/ece)
 -->
+
+## Checklist
+<!--
+Add a checklist of things that are required to be reviewed in order to have the PR approved
+List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply.
+
+List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
```
Comment (simitt) [inline .github/pull_request_template.md]: This line is duplicated.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -21,51 +21,6 @@ or that there are particular issues that you should know about before implementi
...
-we require that you test any code you are adding or modifying.
-Once your changes are ready to submit for review:
-
-1. Sign the Contributor License Agreement
-
-    Please make sure you have signed our [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
-    We are not asking you to assign copyright to us,
-    but to give us the right to distribute your code without restriction.
-    We ask this of all contributors in order to assure our users of the origin and continuing existence of the code.
-    You only need to sign the CLA once.
```
Comment (simitt) [inline CONTRIBUTING.md]: I think this is important information. Can you add it as a comment to the PR template?

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -27,3 +31,41 @@ Major changes require a number of considerations including impact on:
...
+-->
+- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/). 
+- [ ] My code follows the style guidelines of this project
+- [ ] I have rebased my changes on top of the latest master branch
```
Comment (simitt) [inline .github/pull_request_template.md]: We do use a linter in the first CI stage. Should we add a checkbox for ensuring the `check-full` passes?

File: .github/pull_request_template.md (hunk 2 above)
Comment (mdelapenya) [reply .github/pull_request_template.md]: Oopps! 🔥

File: CONTRIBUTING.md (hunk 3 above)
Comment (mdelapenya) [reply CONTRIBUTING.md]: Oh I see. I'll add it alongside the CLA check.

File: CONTRIBUTING.md (hunk 1 above)
Comment (mdelapenya) [reply CONTRIBUTING.md]: What about:
```markdown
### Submitting your changes

Please read our [pull request template](.github/pull_request_template.md), which includes the information we care the most when submitting new changes.
```

File: .github/pull_request_template.md (hunk 4 above)
Comment (mdelapenya) [reply .github/pull_request_template.md]: This template is filled out before reaching the CI. I'd add here actions the sender must do before submitting the PR. Maybe we could explain that in the `My code follows the style guidelines...`. If so, what info would you add there?

Comment (simitt) [reply .github/pull_request_template.md]: Agreed. I was referring to running `make check-full` before creating the PR. So maybe something along the lines of:
_My code follows the style guidelines of this project (run `make check-full` for static code checks and linting)._

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -27,3 +31,42 @@ Major changes require a number of considerations including impact on:
...
+## Related issues
+<!--
+If this PR should close an issue, please add one of the magic keywords
+(e.g. fixes) followed by the issue number. For more info see:
+https://help.github.com/articles/closing-issues-using-keywords/
+Examples:
+- Closes #ISSUE_ID
+- Relates #ISSUE_ID
+- Requires #ISSUE_ID
+- Superseds #ISSUE_ID
```
Comment (axw) [inline .github/pull_request_template.md]:
```suggestion
- Supersedes #ISSUE_ID
```

**Comment by mdelapenya** [comment]
> Still not sure what the value of the feature/enhancement labels are though...

Oh sorry then. I maybe misunderstood your initial feedback. I renamed from feature-request to feature, but now I see your point: moving that to the issue template, right? I can remove it from here then. At the end of the day, this initiative for polishing the template has to make you comfortable with it, so let's remove those labels :)

**Comment by axw** [comment]
@mdelapenya thanks. I'm the only one to bring it up - if anyone else has thoughts on how it can be useful, we should discuss. In general I prefer to start minimal though.

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -27,3 +29,42 @@ Major changes require a number of considerations including impact on:
...
 -->
+
+## Checklist
+<!--
+Add a checklist of things that are required to be reviewed in order to have the PR approved
+List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
+-->
+- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
+<!--
+We are not asking you to assign copyright to us, but to give us the right to distribute your code without restriction. We ask this of all contributors in order to assure our users of the origin and continuing existence of the code. You only need to sign the CLA once.
```
Comment (cachedout) [inline .github/pull_request_template.md]: The CLA specifically does grant Elasticsearch a copyright license. (Section 2 -- Grant of Copyright License).

 Has anybody from legal reviewed this statement? I think we should not attempt to interpret the language of the CLA in a PR template and if we do, we should ensure that legal has signed off on any language.

Comment (mdelapenya) [reply .github/pull_request_template.md]: This is really a good point. We just moved those lines from the CONTIRBUTING guide to here. Maybe it's wise to remove it and delegate the explanation to the CLA itself

Comment (cachedout) [reply .github/pull_request_template.md]: Sounds good. I made a corresponding update to the CONTRIBUTING document so everything is aligned: https://github.com/elastic/apm-server/pull/3292

Comment (mdelapenya) [reply .github/pull_request_template.md]: Won't that PR be conflicting with this one? Here we remove that paragraph

Comment (cachedout) [reply .github/pull_request_template.md]: Ah, I didn't see that. I'll close mine then. Thanks!

**Comment by mdelapenya** [comment]
Is this ready to go? 
image

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -23,48 +23,7 @@ please follow the instructions in the [readme](README.md).
...
-    and in the body provide more details about what your changes do.
-    Also mention the number of the issue where discussion has taken place,
-    eg "Closes #123".
-
-5. Be patient
-
-    We might not be able to review your code as fast as we would like to,
-    but we'll do our best to dedicate it the attention it deserves.
-    Your effort is much appreciated!
+Please read our [pull request template](.github/pull_request_template.md), which includes the information we care the most when submitting new changes.
```
Comment (cachedout) [inline CONTRIBUTING.md]:
```suggestion
Please read our [pull request template](.github/pull_request_template.md), which includes the information we care about the most when submitting new changes.
```

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -3,21 +3,23 @@
 If this is your first contribution, please review and sign our contributor agreement -
 https://www.elastic.co/contributor-agreement.
 
+A few suggestions about filling out this PR:
+
+1. Use a descriptive title for the PR.
+2. If this pull request is work in progress, create a draft PR instead of prefixing the title with WIP.
+3. Please label this PR with at least one of the following labels, depending on the scope of your change:
+- bug fix
+- breaking change
```
Comment (graphaelli) [inline .github/pull_request_template.md]:
```suggestion
- breaking change
- enhancement
```

**Comment by mdelapenya** [comment]
Should I merge this, or is it something the team is doing?

**Comment by graphaelli** [comment]
@mdelapenya go for it

**Pull request merged**